base/linalg/dense.jl

```diff
@@ -169,17 +169,63 @@ kron(a::Vector, b::Vector)=vec(kron(reshape(a,length(a),1),reshape(b,length(b),1
 kron(a::Matrix, b::Vector)=kron(a,reshape(b,length(b),1))
 kron(a::Vector, b::Matrix)=kron(reshape(a,length(a),1),b)
 
-^(A::Matrix, p::Integer) = p < 0 ? inv(A^-p) : Base.power_by_squaring(A,p)
-
-function ^(A::Matrix, p::Number)
+# Matrix power
+^(A::AbstractMatrix, p::Integer) = p < 0 ? Base.power_by_squaring(inv(A),-p) : Base.power_by_squaring(A,p)
+function ^(A::AbstractMatrix, p::Real)
+    # For integer powers, use repeated squaring
     if isinteger(p)
         return A^Integer(real(p))
     end
-    chksquare(A)
-    v, X = eig(A)
-    any(v.<0) && (v = complex(v))
-    Xinv = ishermitian(A) ? X' : inv(X)
-    scale(X, v.^p)*Xinv
+
+    # If possible, use diagonalization
+    if issym(A)
+        return full(Symmetric(A)^p)
+    end
+    if ishermitian(A)
+        return full(Hermitian(A)^p)
+    end
+
+    # General matrices
+    p1 = p - floor(p)
+    if cond(A) >= exp(log(p1/(1 - p1)) / p1)
+        return powm(A,p1) * A^floor(p)
+    else
+        return powm(A,p - ceil(p)) * A^ceil(A)
+    end
+end
+function powm(A::StridedMatrix, p::Real)
+    if abs(p) >= 1
+        ArgumentError("p must be a real number in (-1,1), got $p")
+    end
+
+    # Use Schur decomposition
+    n = chksquare(A)
+    if istriu(A)
+        retmat = full(powm(UpperTriangular(complex(A)),p))
+        d = diag(A)
+    else
+        S,Q,d = schur(complex(A))
+        R = powm(UpperTriangular(S),p)
+        retmat = Q * R * Q'
+    end
+
+    # Check whether the matrix has nonpositive real eigs
+    np_real_eigs = false
+    for i = 1:n
+        if imag(d[i]) < eps() && real(d[i]) <= 0
+            np_real_eigs = true
+            break
+        end
+    end
+    if np_real_eigs
+        warn("Matrix with nonpositive real eigenvalues, a nonprincipal matrix power will be returned.")
+    end
+
+    if isreal(A) && ~np_real_eigs
```

Comment:
If we can be type stable without sacrificing too much convenience, it's better to, even if things aren't performance critical.